### PR TITLE
[ML] Add per allocation and per deployment memory metadata fields to the trained models config

### PR DESCRIPTION
**Pull request by valeriy42**
To improve the required memory estimation of NLP models, this PR introduces two new metadata fields: `per_deployment_memory_bytes` and `per_allocation_memory_bytes`. 

`per_deployment_memory_bytes` is the memory required to load the model in the deployment
`per_allocation_memory_bytes` is the temporary additional memory used during the inference for every allocation.

This PR extends the memory usage estimation logic while ensuring backward compatibility.

In a follow-up PR, I will adjust the assignment planner to use the refined memory usage information.

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/StartTrainedModelDeploymentAction.java [hunk 1]
```diff
@@ -244,6 +252,22 @@ public void setPriority(String priority) {
             this.priority = Priority.fromString(priority);
         }
 
+        public long getPerDeploymentMemoryBytes() {
```
Comment (davidkyle) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/StartTrainedModelDeploymentAction.java:255]: I can't see where these new fields added to the `StartTrainedModelDeploymentAction.Request` object are used

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/ModelPackageConfigTests.java [hunk 2]
```diff
@@ -110,7 +110,8 @@ protected ModelPackageConfig mutateInstance(ModelPackageConfig instance) {
 
     @Override
     protected ModelPackageConfig mutateInstanceForVersion(ModelPackageConfig instance, TransportVersion version) {
-        return instance;
+        ModelPackageConfig.Builder builder = new ModelPackageConfig.Builder(instance);
+        return builder.build();
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/ModelPackageConfigTests.java:114]: You are right. It was a remainder of an intermediate change I had. Removed in ddbeee235da.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/StartTrainedModelDeploymentAction.java (hunk 1 above)
Comment (davidkyle) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/StartTrainedModelDeploymentAction.java:255]: There are 2 classes in the file that have a `getPerDeploymentMemoryBytes()` function. `StartTrainedModelDeploymentAction.Request` on line 255 and `StartTrainedModelDeploymentAction.TaskParams` on line 724.  The `TaskParams functions are used but I don't think the `Request` functions are

Comment (valeriy42) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/StartTrainedModelDeploymentAction.java]: Indeed, you are right. I ended up removing the fields from the `Request` class.

**Comment by valeriy42** [comment]
Thank you for the review @davidkyle . I updated the code as you suggested. It would be great if you could take another pass.

**Review by davidkyle — APPROVED** [review]
LGTM

**Comment by valeriy42** [comment]
retest

**Comment by valeriy42** [comment]
@elasticmachine, run elasticsearch-ci/part-1

**Comment by valeriy42** [comment]
@elasticmachine, run elasticsearch-ci/part-2

**Pull request merged**